File: packages/python/elizaos/features/basic_capabilities/providers/context_bench.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


async def get_context_bench(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    _ = runtime
    _ = state
    meta = getattr(message, "metadata", None)
    bench_ctx = None
    if meta is not None:
        maybe = getattr(meta, "benchmarkContext", None)
        if isinstance(maybe, str) and maybe.strip():
            bench_ctx = maybe.strip()

    if not bench_ctx:
        return ProviderResult(
            text="",
            values={"benchmark_has_context": False},
            data={},
        )

    return ProviderResult(
        text=f"# Benchmark Context\n{bench_ctx}",
        values={"benchmark_has_context": True},
        data={"benchmarkContext": bench_ctx},
    )


context_bench_provider = Provider(
    name="CONTEXT_BENCH",
    description="Benchmark/task context injected by a benchmark harness",
    get=get_context_bench,
    dynamic=True,
)

File: packages/rust/src/features/basic_capabilities/bootstrap/providers/context_bench.rs
//! Benchmark context provider.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("CONTEXT_BENCH"));

pub struct ContextBenchProvider;

#[async_trait]
impl Provider for ContextBenchProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        _runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let bench_ctx = message.metadata.as_ref().and_then(|meta| {
            let crate::types::memory::MemoryMetadata::Custom(v) = meta;
            v.as_object()
                .and_then(|obj| obj.get("benchmarkContext"))
                .and_then(|x| x.as_str())
                .filter(|s| !s.trim().is_empty())
                .map(|s| s.trim().to_string())
        });

        if let Some(ctx) = bench_ctx {
            Ok(ProviderResult {
                text: Some(format!("# Benchmark Context\n{}", ctx)),
                values: Some(
                    [(
                        "benchmark_has_context".to_string(),
                        serde_json::Value::Bool(true),
                    )]
                    .into_iter()
                    .collect(),
                ),
                data: Some(
                    [(
                        "benchmarkContext".to_string(),
                        serde_json::Value::String(ctx),
                    )]
                    .into_iter()
                    .collect(),
                ),
            })
        } else {
            Ok(ProviderResult {
                text: None,
                values: Some(
                    [(
                        "benchmark_has_context".to_string(),
                        serde_json::Value::Bool(false),
                    )]
                    .into_iter()
                    .collect(),
                ),
                data: None,
            })
        }
    }
}
